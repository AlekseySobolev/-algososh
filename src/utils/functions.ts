export const swap = (array: string[] | number[], firstIndex: number, secondIndex: number): string[] | number[] => {
    let temp;
    temp = array[firstIndex];
    array[firstIndex] = array[secondIndex];
    array[secondIndex] = temp;
    return array;
  };

  export const stringSort = (symbolArr: string[]): string[][] => {

    const arrayOfSortedArray: string[][] = [];
    arrayOfSortedArray.push([...symbolArr]);
    
    let i = 0;
    let j = symbolArr.length - 1;

    while(i < j){
      swap(symbolArr, i, j);
      arrayOfSortedArray.push([...symbolArr]);
      i++;
      j--; 
    }
    return arrayOfSortedArray;
  }