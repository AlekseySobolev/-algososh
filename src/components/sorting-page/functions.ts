import { Direction } from "../../types/direction";
import { ElementStates } from "../../types/element-states";
import { swap } from "../../utils/functions";
import { Step } from "./types";

const getIntFromInterval = (min: number, max: number) => {
    return Math.floor(Math.random() * (max - min + 1) + min)
}

export const getRandomArray = (min = 0, max = 100, minLength = 3, maxLength = 17) => {
    const result = new Array(getIntFromInterval(minLength, maxLength)).fill(0);
    return Array.from(new Set(result.map(() => getIntFromInterval(min, max))))
}

export const getColumnState = (index: number, isLastStep: boolean, currerntStep: Step): ElementStates => {
    if ([currerntStep.aIndex, currerntStep.bIndex].includes(index)) {
        return ElementStates.Changing;
    }

    if (currerntStep.sortedIndexes.includes(index) || isLastStep) {
        return ElementStates.Modified;
    }

    return ElementStates.Default;
}

export const bubbleSort = (arrForSort: number[], sortDirect: Direction): Step[] => {

    const sortSteps: Step[] = [];
    let itrNum = 0;
    let isSorted = false;

    do {

        isSorted = false;

        for (let i = 0; i < arrForSort.length - 1 - itrNum; i++) {
            if (sortDirect === Direction.Descending ? arrForSort[i] < arrForSort[i + 1] : arrForSort[i] > arrForSort[i + 1]) {
                swap(arrForSort, i, i + 1);
                isSorted = true;
            }
            sortSteps.push({ currentArray: [...arrForSort], sortedIndexes: [...(sortSteps[sortSteps.length - 1]?.sortedIndexes || [])], aIndex: i, bIndex: i + 1 });
        }

        ++itrNum;
        sortSteps[sortSteps.length - 1]?.sortedIndexes.push(arrForSort.length - itrNum);

    } while (isSorted)

    sortSteps.push({ currentArray: [...arrForSort], sortedIndexes: [...(sortSteps[sortSteps.length - 1]?.sortedIndexes || [])] });

    return sortSteps;
}

export const selectSort = (arrForSort: number[], sortDirect: Direction): Step[] => {

    const sortSteps: Step[] = [];
    let itrNum = 0;
    let isSorted = false;

    do {

        isSorted = false;

        for (let i = 0; i < arrForSort.length - 1 - itrNum; i++) {
            let maxInd = i;
            let j = i + 1;

            if (sortDirect === Direction.Descending ? arrForSort[maxInd] < arrForSort[j] : arrForSort[maxInd] > arrForSort[j]) {
                maxInd = j;
            }

            if (maxInd !== i) {
                swap(arrForSort, i, maxInd);
                isSorted = true;
            }
            sortSteps.push({ currentArray: [...arrForSort], sortedIndexes: [...(sortSteps[sortSteps.length - 1]?.sortedIndexes || [])], aIndex: itrNum, bIndex: itrNum + j});
        }

        ++itrNum;
        sortSteps[sortSteps.length - 1]?.sortedIndexes.push(arrForSort.length - itrNum);

    } while (isSorted)

    sortSteps.push({ currentArray: [...arrForSort], sortedIndexes: [...(sortSteps[sortSteps.length - 1]?.sortedIndexes || [])] });

    return sortSteps;

} 