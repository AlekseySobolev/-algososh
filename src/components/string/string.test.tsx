import { fireEvent, render, screen } from "@testing-library/react";
import '@testing-library/jest-dom/extend-expect';
import React from "react";
import { stringSort } from "../../utils/functions";

describe('string algo tests',
() => {

it('shoud correct even reverse string', ()=>{
    const reversedArray = stringSort(['a','l','g','o']);
    const resultArray = ['o','g','l','a'];
    expect(reversedArray[reversedArray.length-1]).toEqual(resultArray);

})

it('shoud correct odd reverse string', ()=>{
    const reversedArray = stringSort(['a','l','g']);
    const resultArray = ['g','l','a']; 
    expect(reversedArray[reversedArray.length-1]).toEqual(resultArray);
})

it('shoud correct 1 letter', ()=>{
    const singleArray = ['l'];
    expect(stringSort(singleArray)).toMatchSnapshot();
})

it('shoud correct empty string', ()=>{
    const emptyArray = [''];
    expect(stringSort(emptyArray)).toMatchSnapshot();
})
    
}
)