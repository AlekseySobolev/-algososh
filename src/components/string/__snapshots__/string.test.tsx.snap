// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`string algo tests shoud correct 1 letter 1`] = `
Array [
  Array [
    "l",
  ],
]
`;

exports[`string algo tests shoud correct empty string 1`] = `
Array [
  Array [
    "",
  ],
]
`;
